batv-sign.1
.TH "BATV-SIGN" "1" "2014-08-26" "" "BATV-TOOLS"
.SH "NAME"
batv-sign \- Generate a BATV key
.SH "SYNOPSIS"
.nf
\fBbatv-sign\fR [\fIoptions\fR ...] \fIfromaddress\fR
.fi
.SH "DESCRIPTION"
\fBbatv-sign\fR signs \fIfromaddress\fR with BATV and outputs the signed envelope sender address to standard out.
.SH "OPTIONS"
.TP
.BI \-k\ \fIkeyfile\fR
Use the key in \fIkeyfile\fR.  Use batv-keygen(1) to generate a key.  (Default: ~/.batv-key)
.TP
.BI \-K\ \fIkeymapfile\fR
Use the key map file in \fIkeymapfile\fR.  (Default: ~/.batv-keys)
.TP
.BI \-l\ \fIlifetime\fR
Lifetime, in days, of the signature. (Default: 7)
.TP
.BI \-d\ \fIdelimiter\fR
Use \fIdelimiter\fR as the sub address delimiter of the signed BATV address.  \fIdelimiter\fR must be a single character and must be recognized by \fIfromaddress\fR's MTA as a sub address delimiter.  (Default: +)
.SH "SEE ALSO"
batv-validate(1), batv-sendmail(1), batv-milter(8), batv-keygen(1)


NEWS
v0.8 (2016-03-13)
  * Fix bug where an invalid username in the config would cause SIGABRT.
  * Eliminate batv-keygen's dependency on OpenSSL.
  * Minor typo fixes and documentation improvements.

v0.7 (2015-05-10)
  * Detect and remove stale socket files when starting up. This is
    helpful if the system shuts down uncleanly and the socket file is
    stored on a persistent filesystem.
  * Ignore trailing whitespace on lines when parsing config files.
  * Eliminate dependency on OpenSSL.
  * Minor documentation improvements.

v0.6 (2014-12-16)
  * Add man pages.
  * Add support for $BATV_KEY_MAP_FILE in batv-sendmail for specifying
    key map file.
  * Fix build errors on FreeBSD.
  * Speed up I/O in batv-validate.
  * Update documentation to say that IPv6 is supported.
  * Improve error handling when loading key files.
  * Makefile: add DESTDIR support, also install batv-keygen.
  * Minor bug fixes.

v0.5 (2013-12-17)
  (See RELEASE_NOTES-0.5 for important details.)
  * Enhance 'X-Batv-Status' header to specify reason for invalidity:
     - valid
     - invalid, missing
     - invalid, bad-signature
     - invalid, multiple-rcpt
  * Add bounce detection logic to batv-milter itself, so it can add a
    'X-Batv-Status: invalid, missing' header if a bounce is missing a
    BATV signature.
  * 'batv-filter -f' now adds a 'X-Batv-Status: invalid, missing' header
    if message did not contain a BATV signature.
  * Add 'on-invalid' option to batv-milter to specify what the milter
    should do with invalid bounces ('accept', 'reject', or 'tempfail').
    Default is 'accept', which was the behavior in previous releases.
  * Improve documentation.

v0.4 (2013-05-28)
  * Fix a build bug due to missing #includes.

v0.3 (2013-04-27)
  * Add batv-keygen script for generating BATV key.
  * Prevent a spurious warning from libmilter.
  * Add initial documentation.

v0.2 (2013-03-13)
  * Fix bug in batv-sendmail.
  * batv-milter: set X-Batv-Delivered-To header.

v0.1 (2013-02-27)
  * Initial release.


batv-sign.cpp
/*
 * Copyright 2013 Andrew Ayer
 *
 * This file is part of batv-tools.
 *
 * batv-tools is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * batv-tools is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with batv-tools.  If not, see <http://www.gnu.org/licenses/>.
 *
 * Additional permission under GNU GPL version 3 section 7:
 *
 * If you modify the Program, or any covered work, by linking or
 * combining it with the OpenSSL project's OpenSSL library (or a
 * modified version of that library), containing parts covered by the
 * terms of the OpenSSL or SSLeay licenses, the licensors of the Program
 * grant you additional permission to convey the resulting work.
 * Corresponding Source for a non-source form of such a combination
 * shall include the source code for the parts of OpenSSL used as well
 * as that of the covered work.
 */

#include "prvs.hpp"
#include "key.hpp"
#include "common.hpp"
#include "address.hpp"
#include <iostream>
#include <fstream>
#include <unistd.h>
#include <errno.h>
#include <cstring>
#include <cstdlib>
#include <string>
#include <string.h>

using namespace batv;

namespace {
	void print_usage (const char* argv0)
	{
		std::clog << "Usage: " << argv0 << " [OPTIONS...] FROM_ADDRESS" << std::endl;
		std::clog << "Options:" << std::endl;
		std::clog << " -k KEY_FILE        -- path to key file (default: ~/.batv-key)" << std::endl;
		std::clog << " -K KEY_MAP_FILE    -- path to key map file (default: ~/.batv-keys)" << std::endl;
		std::clog << " -l LIFETIME        -- lifetime, in days, of BATV address (default: 7)" << std::endl;
		std::clog << " -d SUB_ADDR_DELIM  -- sub address delimiter (default: +)" << std::endl;
	}
}

int main (int argc, char** argv)
try {
	char		sub_address_delimiter = '+';
	unsigned int	address_lifetime = 7;
	Key		key;
	std::string	key_file;
	Key_map		key_map;
	std::string	key_map_file;

	int		flag;
	while ((flag = getopt(argc, argv, "k:K:l:d:")) != -1) {
		switch (flag) {
		case 'k':
			key_file = optarg;
			break;
		case 'K':
			key_map_file = optarg;
			break;
		case 'l':
			address_lifetime = std::atoi(optarg);
			break;
		case 'd':
			if (std::strlen(optarg) != 1) {
				std::clog << argv[0] << ": sub address delimiter (as specified by -d) must be exactly one character" << std::endl;
				return 1;
			}
			sub_address_delimiter = optarg[0];
			break;
		default:
			print_usage(argv[0]);
			return 2;
		}
	}

	if (argc - optind != 1) {
		print_usage(argv[0]);
		return 2;
	}

	if (address_lifetime < 1 || address_lifetime > 999) {
		std::clog << argv[0] << ": address lifetime (as specified by -l) must be between 1 and 999, inclusive" << std::endl;
		return 1;
	}

	// Load the key
	check_personal_key_path(key_file, ".batv-key");
	check_personal_key_path(key_map_file, ".batv-keys");

	if (key_file.empty() && key_map_file.empty()) {
		std::clog << argv[0] << ": Neither ~/.batv-key nor ~/.batv-keys exist." << std::endl;
		std::clog << "Please create one and/or the other or specify alternative paths using -k or -K" << std::endl;
		return 1;
	}

	// Load the key and key map
	if (!key_file.empty()) {
		load_key(key, key_file);
	}
	if (!key_map_file.empty()) {
		std::ifstream	key_map_in(key_map_file.c_str());
		load_key_map(key_map, key_map_in);
	}
	
	// Determine what key to use to sign this message
	const Key*		use_key = get_key(key_map, argv[optind], !key.empty() ? &key : NULL);
	if (!use_key) {
		std::clog << argv[0] << ": " << argv[optind] << ": No key available for this sender" << std::endl;
		return 1;
	}

	// Generate the BATV address
	Email_address		from_address;
	from_address.parse(argv[optind]);
	if (from_address.domain.empty()) {
		std::clog << argv[0] << ": " << argv[optind] << ": Address is missing domain name" << std::endl;
		return 1;
	}

	std::cout << prvs_generate(from_address, address_lifetime, *use_key).make_string(sub_address_delimiter) << std::endl;
	return 0;

} catch (const Initialization_error& e) {
	std::clog << argv[0] << ": " << e.message << std::endl;
	return 1;
}

